### ISSUE TITLE
Implementation inceptionResnetV2 backbone

### ISSUE THREAD

**Issue by Maxfashko** [fizyr/keras-retinanet#799]
Hi fizyr Team, I am trying to implement backbone "inceptionResnetV2".
1) I added support for a new backbone https://github.com/fizyr/keras-retinanet/blob/master/keras_retinanet/models/__init__.py#L59

```
elif 'inceptionResnetV2' in backbone_name:
        from .inceptionresnetv2 import InceptionResNetV2Backbone as b
```

2) I implemented inceptionResnetV2.py:

```
from keras.applications.inception_resnet_v2 import InceptionResNetV2, preprocess_input, decode_predictions


class InceptionResNetV2Backbone(Backbone):
    def __init__(self, backbone):
        super(InceptionResNetV2Backbone, self).__init__(backbone)
        self.custom_objects.update(keras_resnet.custom_objects)

    def retinanet(self, *args, **kwargs):
        return resnet_retinanet(*args, backbone=self.backbone, **kwargs)

    def download_imagenet(self):
        filename = 'inception_resnet_v2_weights_tf_dim_ordering_tf_kernels_notop.h5'
        resource = 'https://github.com/fchollet/deep-learning-models/releases/download/v0.7/{}'.format(filename)
        checksum = 'd19885ff4a710c122648d3b5c3b684e4'

        return get_file(
            filename,
            resource,
            cache_subdir='models',
            md5_hash=checksum
        )

    def validate(self):
        allowed_backbones = ['inceptionResnetV2']
        backbone = self.backbone.split('_')[0]

        if backbone not in allowed_backbones:
            raise ValueError('Backbone (\'{}\') not in allowed backbones ({}).'.format(backbone, allowed_backbones))

    def preprocess_image(self, inputs):
        return preprocess_image(inputs, mode='caffe')


def resnet_retinanet(num_classes, backbone='inceptionResnetV2', inputs=None, modifier=None, **kwargs):
    if inputs is None:
        if keras.backend.image_data_format() == 'channels_first':
            inputs = keras.layers.Input(shape=(3, None, None))
        else:
            inputs = keras.layers.Input(shape=(None, None, 3))

    if backbone == 'inceptionResnetV2':
        from keras.models import Model

        model = InceptionResNetV2(input_tensor=inputs)
        model = Model(inputs=model.input, outputs=model.get_layer('avg_pool').output)
        c3 = model.get_layer('activation_5').output
        c4 = model.get_layer('activation_76').output
        c5 = model.get_layer('activation_163').output

        outputs = (c3, c4, c5)

    else:
        raise ValueError('Backbone (\'{}\') is invalid.'.format(backbone))

    if modifier:
        resnet = modifier(resnet)

    return retinanet.retinanet(inputs=inputs, num_classes=num_classes, backbone_layers=outputs, **kwargs)


def inceptionresnetv2_retinanet(num_classes, inputs=None, **kwargs):
    return resnet_retinanet(num_classes=num_classes, backbone='inceptionResnetV2', inputs=inputs, **kwargs)
```

I chose the names of the layers that precede the global pooling. Now, when I start the learning process, the mAP remains unchanged.

```
Total params: 36,817,837
Trainable params: 36,780,013
Non-trainable params: 37,824
__________________________________________________________________________________________________

Epoch 1/50
200/200 [==============================] - 89s 443ms/step - loss: 3.5791 - regression_loss: 2.6202 - classification_loss: 0.9590
mAP: 0.0000

Epoch 00001: saving model to ./snapshots/inceptionResnetV2_pascal_01.h5
Epoch 2/50
200/200 [==============================] - 71s 354ms/step - loss: 3.3621 - regression_loss: 2.5347 - classification_loss: 0.8274
mAP: 0.0000

Epoch 00002: saving model to ./snapshots/inceptionResnetV2_pascal_02.h5
Epoch 3/50
200/200 [==============================] - 71s 355ms/step - loss: 3.3349 - regression_loss: 2.5394 - classification_loss: 0.7955
mAP: 0.0000

Epoch 00003: saving model to ./snapshots/inceptionResnetV2_pascal_03.h5
Epoch 4/50
200/200 [==============================] - 71s 357ms/step - loss: 3.3361 - regression_loss: 2.5274 - classification_loss: 0.8088
mAP: 0.0000
```

I have two suggestions:
1) I chose the wrong layers for use in FPN
2) I missed something else in the implementation, and I need to add something.

I would be grateful for any help.

**Comment by hgaiser** [comment]
Hi, cool, I'm curious to see how it develops :)

It seems you've only trained for 4 epochs, with step size 200 per epoch. It could be that you simply need to train longer, your loss does seem to go down. Something you could look into: if you create a "normal" inceptionresnetv2 classification network and preprocess an image (the same way as how you preprocess currently), resize it to 224x224, does it correctly classify the object?

**Comment by hgaiser** [comment]
Any update on this ?

**Comment by Maxfashko** [comment]
@hgaiser I am very busy at work. I did not deal with this problem.

**Comment by hgaiser** [comment]
I understand :+1: 

Let me know if there is something to share.

**Comment by hgaiser** [comment]
I'll close this for now as it is not an issue with the code.
